akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -346,9 +349,12 @@ final class Split[T](decision: Split.SplitDecision, p: T ⇒ Boolean) extends Gr
             closeThis(handler, elem)
             handOver(handler)
           } else {
-            // Drain into the void
-            if (substreamCancelled) pull(in)
-            else substreamSource.push(elem)
+            if (substreamCancelled) {
```

Comment:
is it actually what you want? I though you want to cancel the upstream and complete possible substreams if the master stream is cancelled. Currently it cancels the upstream and everything when a substream is cancelled. But AFAIK substreams *are* the entitiy streams already, so if someone gets to cancel them that means they are properly consumed.